### ISSUE TITLE
[css-grid] axis names

### ISSUE THREAD

**Issue by chharvey** [w3c/csswg-drafts#1299]
https://www.w3.org/TR/css-grid-1/#grid-concepts

From the spec,

> There are two sets of grid lines: one set defining columns that run along the block axis (the column axis), and an orthogonal set defining rows along the inline axis (the row axis).

This seems backwards to me. The set of grid lines containing the columns should be called the **column axis**. This axis, in top-bottom writing modes, would be horizontal—parallel to the window's x-axis and to the inline axis. The set of rows should be called the **row axis**, which, in top-bottom writing modes, would be vertical—parallel to the y- and block axes.

From a mathematical point of view, the name of the axis indicates the values that lie on that axis. E.g. the "x-axis" is the axis on which x-values increase. So it seems that naturally, the axis on which column numbers increase should be called the **column axis** (horizontal), and the axis on which row numbers increase (the vertical axis) should be called the **row axis**.

I'm not sure if switching these terms is a good idea, however, given the precedent and the status of the CSS Grid Layout Module Level 1 document (currently, Candidate Recommendation). However I wanted to express my concerns anyway, in case my viewpoint hadn't been considered.

**Comment by SelenIT** [comment]
Block and inline axes are defined in [css-writing-modes-3](https://drafts.csswg.org/css-writing-modes-3/#abstract-axes). The whole point of the flow-relative directions (as opposed to physical directions) is that in vertical writing modes the block axis becomes horizontal, and the inline axis becomes vertical. The columns and the rows of the grid, respectively, do the same. So I see no contradiction in the current definition.

**Comment by Loirooriol** [comment]
@SelenIT But the problem seems that the names should be the opposite. In an horizontal writing mode, the block axis is vertical, and the spec says this is the column axis. But columns are vertical in an horizontal writing mode, so the column axis should be the one which intersects each column: the horizontal one.

**Comment by SelenIT** [comment]
Oh, I see now. It seemed so natural for me to interpret the column axis as a direction where the columns _go_ (as it is currently in the spec) that it took two comments just to make me see the other possibility to interpret it:)

Now I can understand the logic behind calling 'the column axis' the axis perpendicular to the colums direction, but I'm still not convinced that this would be 'more natural'. In everyday use, 'the column direction' is clearly the direction where _the columns themselves_ go, not where _their numbers_ go. And having 'column axis' perpendicular to the 'column direction' seems rather surprising (for me, at least). With mathematical x-axis, there is no such problem, we usually think of x-values not as of vertical lines, so 'the x-value direction' seems to be clearly horizontal.

Maybe it would be better to remove the terms 'row axis' and 'column axis' at all, leaving just block and inline axes which doesn't have that ambiguity?

**Comment by rachelandrew** [comment]
I've encountered people confused by the Column and Row terminology as well. I'd generally prefer to teach Block and Inline to people, the multiple names seems more confusing than helpful.

**Comment by FremyCompany** [comment]
Make a pull request to update the terminology to always use one?

**Comment by chharvey** [comment]
@SelenIT — I like the idea of using **block-axis** and **inline-axis** to be consistent with the writing mode and direction. I haven't tested, but I assume in a *vertical writing mode* that the "column axis" (as it is currently called) would be horizontal, and "row axis" (current definition) would be vertical? In that case it does indeed seem counter-intuitive to call them "columns" and "rows".

So this poses the question of whether naming all the grid properties (e.g., `grid-template-columns`) is relative to writing mode. But I think if we convince ourselves that "column" and "row" need not directly correspond do "vertical" and "horizontal", respectively, then we can let the property names slide.

I still think that the current naming of the axes should be revisited. If "column-axis" and "row-axis" can't be switched, I'm in favor of "block-axis" and "inline-axis."

In fact, while we're at it, it would also be nice if "cross-axis" and "main-axis" (in the Flexbox layout) were also respectively changed to "block-axis" and "inline-axis," for consistency's sake.

**Comment by Loirooriol** [comment]
@chharvey I wouldn't change the terminology in flexbox. For example, in horizontal writing mode, the block axis is defined to be vertical, but you can still have either a vertical or horizontal main axis via flex-direction. I guess it would be possible to redefine block and inline axis to mean different things depending on the formatting context, but changing the terminology would be confusing because flex layout does not have blocks and inlines, all flex items are blockified.

But 'row axis' and 'column axis' don't seem to be relevant in CSS Grids, they are only used in a sentence which references both and in an example. I don't see any problem in removing them. Otherwise they should be swapped for consistency.

    ┌————————————————————————————→  Inline axis
    │  ┌─Block─────────────────┐
    │  │ ┌─Inline─┐ ┌─Inline─┐ │
    │  │ └────────┘ └────────┘ │
    │  └───────────────────────┘
    │  ┌─Block─────────────────┐
    │  └───────────────────────┘
    ↓ Block axis

    ┌—————————————————————————————————→  Row axis (current definition)
    │        ┌─Column─┐  ┌─Column─┐
    │  ┌─Row─┼────────┼──┼────────┼─┐
    │  └─────┼────────┼──┼────────┼─┘
    │  ┌─Row─┼────────┼──┼────────┼─┐
    │  └─────┼────────┼──┼────────┼─┘
    │        └────────┘  └────────┘
    ↓ Column axis (current definition)

**Comment by javifernandez** [comment]
I'm not sure row/column terminology is that useless for Grid, really. Many of the Grid CSS properties uses these concepts (grid-template-{columns, rows}, grid-auto-{rows, columns}, grid-auto-flow: [row | column],  ...) 

Additionally, row and column axis are terms specific to the grid container, while grid items may have orthogonal inline/block axis. This is specially useful when implementing the alignment logic defined in the CSS Box Alignment specification.

https://drafts.csswg.org/css-align-3/#align-grid
> align-self Axis 	The grid’s column axis (block axis). 

Note that align-self is a grid item's property, but it operates along the **column axis**, which certainly is the block axis, but not respecting the grid item's flow direction, but grid container's one.

Even more, all this concepts are specially important when implementing Baseline Alignment, which
in the case of grid it can operate along both axis, aligning the items between themselves in both
dimensions. 

In summary, I don't see such a problem on having this grid specific terminology. It may be redundant
in some cases, but it helps to understand the behavior of grid in some other cases, probably the most
complex ones. I think it's quite similar to what Flexbox's cross and main axis provide, actually.

**Comment by tabatkins** [comment]
Reasonable arguments can be made for making the "column axis" (and "row axis") go in either direction:

* "column axis" should be oriented the same way as columns themselves (vertical in English)
* "column axis" should be the direction that a column grows (vertical)
* "column axis" should be the direction that you add new columns in / number the columns (horizontal)

(2) and (3) from above *both* correspond to how block/inlines work - as you add more content to a block/inline, it grows vertical/horizontally, but if you add more blocks/inlines, they *also* get added vertically/horizontally, so you can't use block/inline as a pattern to follow here.

Ultimately, it was "column axis should put you in mind of a column" that decided us on the current definition.

**Comment by Loirooriol** [comment]
Personally I think that was a bad decision, but the reasoning is interesting. Maybe you can include it as a note in the spec to avoid confusion?

**Comment by fantasai** [comment]
Added note, as requested. Let me know if it needs further tweaking, closing issue for now~

**Comment by SelenIT** [comment]
In her recent article [Understanding Logical Properties And Values](https://www.smashingmagazine.com/2018/03/understanding-logical-properties-values/), @rachelandrew makes a very important point for naming the axes as proposed here:

> In CSS Grid Layout the Block Axis is also referred to as the Row Axis, which is why the Block Axis properties are `grid-row-start` and `grid-row-end`.
> ...the Inline Axis is the Column Axis with the properties `grid-column-start` and `grid-column-end`.

If renamed this way, the names of the axes would be consistent with the Logical Properties/axes names (as Rachel explains further in this article). Maybe it would make sense to reopen this issue?

**Comment by fantasai** [comment]
Reopening per @SelenIT’s comment.
